EG-2007: add internal documentation for BN operations

bpaunescu opened the pull request.

tubibuto committed:
EG-2292 Introduced MembershipState and supporting tests (#6352)

tubibuto committed:
EG-2293 Membership Contract for Business Networks (#6355)

* EG-2293 Implemented contract for membership state and its test coverage

* EG-2293 Added transaction signing related contract verification logic

* EG-2293 Addressed redundant test helper method comments

tubibuto committed:
EG-2294 Create Business Network flow implementation (#6363)

* EG-2293 Implemented contract for membership state and its test coverage

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Database service KDocs

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Addressed comments regarding network ID and notary flow arguments and rebased successfully

* EG-2294 Created separate DatabaseService method to check whether BN with given ID exists

tubibuto committed:
EG-2294 Request Membership flow implementation (#6364)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented RequestMembershipFlow and added unit test coverage

* EG-2294 Added more refined error handling

* EG-2294 Added KDocs to custom exceptions

* EG-2294 added notary optional argument to RequestMembershipFlow

* EG-2294 Updated KDocs for RequestMembershipFlow

tubibuto committed:
EG-2294 Request Membership flow implementation (#6364)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented RequestMembershipFlow and added unit test coverage

* EG-2294 Added more refined error handling

* EG-2294 Added KDocs to custom exceptions

* EG-2294 added notary optional argument to RequestMembershipFlow

* EG-2294 Updated KDocs for RequestMembershipFlow

tubibuto committed:
EG-2294 Activate Membership flow implementation (#6369)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented RequestMembershipFlow and added unit test coverage

* EG-2294 Implemented RequestMembershipFlow and added unit test coverage

* EG-2294 Implemented ActivateMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary argument to ActivateMembershipFlow

* EG-2294 Adapted output state participants to match growing set of observers

* EG-2294 Fixed missing test dependency issue

tubibuto committed:
EG-2294 Suspend Membership flow implementation (#6370)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Implemented ActivateMembershipFlow and added unit test coverage

* EG-2294 Fixed missing test dependency issue

* EG-2294 Implemented SuspendMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary argument

tubibuto committed:
EG-2294 Suspend Membership flow implementation (#6370)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Implemented ActivateMembershipFlow and added unit test coverage

* EG-2294 Fixed missing test dependency issue

* EG-2294 Implemented SuspendMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary argument

tubibuto committed:
EG-2294 Revoke Membership flow implementation (#6371)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed missing test dependency issue

* EG-2294 Implemented SuspendMembershipFlow and added unit test coverage

* EG-2294 Implemented SuspendMembershipFlow and added unit test coverage

* EG-2294 Implemented RevokeMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary flow argument

* EG-2294 Added assertion messages in loops

tubibuto committed:
EG-2294 Revoke Membership flow implementation (#6371)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed missing test dependency issue

* EG-2294 Implemented SuspendMembershipFlow and added unit test coverage

* EG-2294 Implemented SuspendMembershipFlow and added unit test coverage

* EG-2294 Implemented RevokeMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary flow argument

* EG-2294 Added assertion messages in loops

tubibuto committed:
EG-2754, EG-2755 Modify Roles flow implementation (#6405)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Implemented RequestMembershipFlow and added unit test coverage

* EG-2294 Implemented ActivateMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary argument to ActivateMembershipFlow

* EG-2294 Fixed missing test dependency issue

* EG-2294 Implemented RevokeMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary flow argument

* EG-2754, EG-2755 Implemented ModifyRolesFlow, supporting states and contracts and tests

* EG-2754, EG-2755 Renamed helper method to better reflect its purpose

* EG-2754, EG-2755 Addressed PR comments and fixed failing tests

tubibuto committed:
EG-2754, EG-2755 Modify Roles flow implementation (#6405)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Implemented RequestMembershipFlow and added unit test coverage

* EG-2294 Implemented ActivateMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary argument to ActivateMembershipFlow

* EG-2294 Fixed missing test dependency issue

* EG-2294 Implemented RevokeMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary flow argument

* EG-2754, EG-2755 Implemented ModifyRolesFlow, supporting states and contracts and tests

* EG-2754, EG-2755 Renamed helper method to better reflect its purpose

* EG-2754, EG-2755 Addressed PR comments and fixed failing tests

tubibuto committed:
EG-2754, EG-2755 Modify Additional Identity flow implementation (#6417)

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Fixed failing tests due to missing test dependency

* EG-2294 Implemented RequestMembershipFlow and added unit test coverage

* EG-2294 Implemented ActivateMembershipFlow and added unit test coverage

* EG-2294 Implemented RevokeMembershipFlow and added unit test coverage

* EG-2294 Addressed PR comments and added optional notary argument to ActivateMembershipFlow

* EG-2294 Addressed PR comments and added optional notary flow argument

* EG-2294 Fixed missing test dependency issue

* EG-2294 Implemented CreateBusinessNetworkFlow and added unit test coverage

* EG-2294 Implemented RequestMembershipFlow and added unit test coverage

* EG-2294 Implemented ActivateMembershipFlow and added unit test coverage

* EG-2294 Implemented SuspendMembershipFlow and added unit test coverage

* EG-2294 Implemented RevokeMembershipFlow and added unit test coverage

* EG-2294 added notary optional argument to RequestMembershipFlow

* EG-2294 Addressed PR comments and added optional notary argument to ActivateMembershipFlow

* EG-2294 Addressed PR comments and added optional notary flow argument

* EG-2294 Fixed missing test dependency issue

* EG-2754, EG-2755 Implemented ModifyRolesFlow, supporting states and contracts and tests

* EG-2754, EG-2755 Introduced additional identity to membership and adapted existing codebase to it

* EG-2754, EG-2755 Implemented ModifyAdditionalIdentityFlow and added test support for it

* EG-2754, EG-2755 Added supporting KDocs to new state fields and flows

* EG-2754, EG-2755 Finished Membership Contract logic for additional entity modification transaction

* EG-2754, EG-2755 Added additional parameter argument to CreateBusinessNetworkFlow and RequestMembershipFlow

* EG-2754, EG-2755 Fixed failing test

* Rebase and renamed additional identity to business identity

tubibuto committed:
EG-2754, EG-2755 Group State and Contract for Business Network (#6442)

* EG-2754, EG-2755 Created group state, contract logic and added test coverage

* EG-2754, EG-2755 Added moar KDocs

tubibuto committed:
EG-2754, EG-2755 Group Management flows implementation (#6444)

* EG-2754, EG-2755 Created group state, contract logic and added test coverage

* EG-2754, EG-2755 Implemented group management flow and adapted existing membership management flows and tests

* EG-2754, EG-2755 Fixed minor error with members who need to have participant list synced

* EG-2754, EG-2755 Implemented test coverage for group management flows (#6445)

* EG-2754, EG-2755 Addressed PR comments

* EG-2754, EG-2755 Fixed detekt issues

bpaunescu committed:
EG-2007: add internal documentation for BN operations

bpaunescu committed:
EG-2007: add internal documentation for BN operations

tubibuto commented on business-networks/README.md:
```diff
@@ -0,0 +1,169 @@
+# Business networks and memberships
+
+The Corda platform extension for creating and managing business networks allows a node operator to define and create a logical network based on a set of common CorDApps as well as a shared
```
nit: CorDapps

tubibuto replied:
Do we want to call this platform extension since it is a CorDapp ?

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+**Flow arguments:**
+
+- ```networkId``` Custom ID to be given to the new Business Network. If not specified, a randomly selected one will be used
+- ```businessIdentity``` Custom business identity to be given to membership
````
Maybe it's worth stressing out this is optional argument

bpaunescu replied:
Will do.

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+*Example*:
+```kotlin
+val myIdentity: BNIdentity = createBusinessNetworkIdentity() // create an instance of a class implementing [BNIdentity]
+val notary = serviceHub.networkMapCache.notaryIdentities.first())
````
nit: Extra closing parentheses

tubibuto commented on business-networks/README.md:
```diff
@@ -0,0 +1,169 @@
+val notary = serviceHub.networkMapCache.notaryIdentities.first())
+
+CordaRPCClient(rpcAddress).start(user.userName, user.password).use {
+    it.proxy.startFlow(::CreateBusinessNetworkFlow, "MyBusinessNetwork", myIdentity, "1", "X-men-and-women", notary)
```
I've never tried this input, does it really work ?

For instance I don't know whether "1" will be correctly parsed as `UniqueIdentifier`

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+- ```groupName``` Optional name to be given to the Business Network group
+- ```notary``` Identity of the notary to be used for transactions notarisation. If not specified, first one from the whitelist will be used
+
+*Example*:
````
We could also give a basic node shell example of creating Business Network without any optional arguments - something like
```
flow start CreateBusinessNetworkFlow
```

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+## On-board a new member
+
+Joining a business network is a 2 step process. First, the Corda node wishing to join must run the ```RequestMembershipFlow``` either from the node shell or any other RPC client.
+As a result of a successful run, a membership is created with a *PENDING* status. Until activated by an authorised party (a business network operator for instance) the newly generated
````
Maybe mention that it is stored on ledgers of all authorised members ?

bpaunescu replied:
Perhaps rephrase to something like "a membership is created with a PENDING status and all authorised members will have visibility over it for further operations".

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+
+- ```authorisedParty``` Identity of authorised member from whom membership activation is requested
+- ```networkId``` ID of the Business Network that potential new member wants to join
+- ```businessIdentity``` Custom business identity to be given to membership
````
Maybe worth mentioning this is an optional argument

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+- ```businessIdentity``` Custom business identity to be given to membership
+- ```notary``` Identity of the notary to be used for transactions notarisation. If not specified, first one from the whitelist will be used
+
+*Example*:
````
We could also give a basic node shell example with no optional arguments:
```
flow start RequestMembershipFlow authorisedParty: BestBNO, networkId: MyBusinessNetwork, businessIdentity: null, notary: null
```

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+```
+
+To finalise the on-boarding process, an authorised party needs to run the ```ActivateMembershipFlow```. This will update the targeted membership status from *PENDING* to *ACTIVE*.
+Activation requires approval from **all** authorised parties in the network.
````
Maybe we could specify that approval means collecting signatures from all authorised parties in the network

bpaunescu replied:
Yes, the statement is a bit vague and can be interpreted as the other authorised parties are required to do extra operations with manual intervention.

bpaunescu replied:
Also: Whoever activates membership needs to assign a group to the newly onboarded member.

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+- ```membershipId``` ID of the membership to be activated
+- ```notary``` Identity of the notary to be used for transactions notarisation. If not specified, first one from the whitelist will be used
+
+*Example*:
````
Potential node shell example:
```
flow start ActivateMembershipFlow membershipId: 123456, notary: null
```

tubibuto commented on business-networks/README.md:
```diff
@@ -0,0 +1,169 @@
+val notary = serviceHub.networkMapCache.notaryIdentities.first())
+
+CordaRPCClient(rpcAddress).start(user.userName, user.password).use {
+    it.proxy.startFlow(::ActivateMembershipFlow, 123456, notary)
```
Will `123456` correctly parse as `UniqueIdentifier` ?

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+
+```kotlin
+val bnService = serviceHub.cordaService(DatabaseService::class.java)
+val updatedIdentity: BNIdentity = updateBusinessIdentity(bnService.getMembership("MyBusinessNetwork", partyToBeUpdated))
````
What does `updateBusinessIdentity()` method mean ?

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+There are several ways in which a member's information can be updated, not including network operations such as membership suspension or revocation. These attributes which can be amended are:
+business network identity, membership list or group, and roles.
+
+To update a member's business identity attribute, one of the authorised network parties needs to run the ```ModifyBusinessIdentityFlow``` which then requires all network members with
````
Maybe it is worth clarifying what is the business identity, maybe even mentioning with what states users are dealing (`MembershipState` and `GroupState`)

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+```
+
+Updating a member's roles and permissions in the business network is done in a similar fashion by using the ```ModifyRolesFlow```. Depending on the proposed changes, the updated member may become
+an authorised member, in which case they will be added to all membership lists in order to be notified of any future business network changes.
````
Members will not be automatically added to all membership lists, user must do that before granting it authorised permission. This should be added as an important note to the docs

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+*Example*:
+
+```kotlin
+val roles = setOf(BNORole) // assign full permissions to member
````
```suggestion
val roles = setOf(BNORole()) // assign full permissions to member
```

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+}
+```
+
+To manage the membership lists or groups, one of the authorised members of the network can use ```DeleteGroupFlow``` or ```ModifyGroupFlow```.
````
What about `CreateGroupFlow` ?

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+
+To manage the membership lists or groups, one of the authorised members of the network can use ```DeleteGroupFlow``` or ```ModifyGroupFlow```.
+
+**DeleteGroupFlow arguments**:
+
+- ```groupId``` ID of group to be deleted
+- ```notary``` Identity of the notary to be used for transactions notarisation. If not specified, first one from the whitelist will be used
````
Will the example come as part of the following commit ?

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+
+The ```ModifyGroupFlow``` can update the name of a group and/or its list of members.
+
+***ModifyGroupFlow arguments**:
````
nit: Extra `*` at the beginning

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+***ModifyGroupFlow arguments**:
+
+- ```groupId``` ID of group to be modified
+- ```name``` New name of modified group
+- ```participants``` New participants of modified group
````
Maybe worth mentioning that at least one of this 2 arguments should be provided

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+
+```kotlin
+val bnService = serviceHub.cordaService(DatabaseService::class.java)
+val participantsList = getBusinessNetworkGroup("1").state.data.participants
````
`getBusinessNetworkGroup()` takes `UniqueIdentifier` and not `String`

bpaunescu replied:
it's bnService.getBusinessNetworkGroup(uniqueIdentifier).

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+```kotlin
+val bnService = serviceHub.cordaService(DatabaseService::class.java)
+val participantsList = getBusinessNetworkGroup("1").state.data.participants
+val newParticipantsList = removeMember(getParty("Wolverine"), participantsList)
````
`participants` is a set of `UniqueIdentifier`s (memberships' linear IDs)

tubibuto commented on business-networks/README.md:
```diff
@@ -0,0 +1,169 @@
+val notary = serviceHub.networkMapCache.notaryIdentities.first())
+
+CordaRPCClient(rpcAddress).start(user.userName, user.password).use {
+    it.proxy.startFlow(::ModifyGroupFlow, "1"", "New Mutants", newParticipantsList, notary)
```
nit: Extra `"` char

tubibuto commented on business-networks/README.md:
````diff
@@ -0,0 +1,169 @@
+}
+```
+
+## Suspend or revoke a membership
````
This will come as part of the next commit or next PR ?

tubibuto reviewed (commented):
Mostly correct, the majority of the comments revolve around making docs bit richer to prevent QA raising unnecessary bug tickets caused by misunderstanding the solution.

One more thing - Maybe it would be worth mentioning which administrative permission should member have to initiate each of the flows (i.e. `CAN_ACTIVATE_MEMBERSHIP`)

**Edit**: Maybe it would be worth having references section and put [primitives doc flows](https://github.com/corda/platform-eng-design/blob/master/extensions/biznet/primitives-design.md#flows) there

bpaunescu committed:
EG-2007: add internal documentation for BN operations

bpaunescu committed:
EG-2007: add internal documentation for BN operations

bpaunescu committed:
EG-2007: rebase

benpesterr3 commented on business-networks/README.md:
`````diff
@@ -0,0 +1,235 @@
+}
+```
+
+To manage the membership lists or groups, one of the authorised members of the network can use ```CreateGroupFlow```, ````DeleteGroupFlow``` and ```ModifyGroupFlow```.
`````
```suggestion
To manage the membership lists or groups, one of the authorised members of the network can use ```CreateGroupFlow```, ```DeleteGroupFlow``` and ```ModifyGroupFlow```.
```

benpesterr3 commented on business-networks/README.md:
````diff
@@ -0,0 +1,235 @@
+
+## Create a business network
+
+Either from the node shell or from an RPC client, simply run ```CreateBusinessNetworkFlow```. This will self issue a membership with an exhaustive permissions set that will
````
```suggestion
Either from the node shell or from an RPC client, run ```CreateBusinessNetworkFlow```. This will self issue a membership with an exhaustive permissions set that will
```

benpesterr3 replied:
```suggestion
Either from the node shell or from an RPC client, simply run `CreateBusinessNetworkFlow`. This will self issue a membership with an exhaustive permissions set that will
```

benpesterr3 commented on business-networks/README.md:
```diff
@@ -0,0 +1,235 @@
+
+In a business network, there is at least one *authorised member*. This member has sufficient permissions to execute management operations over the network and its members.
+
+# Create a business network and manage it
```
```suggestion
## Creating and managing a business network
```

benpesterr3 commented on business-networks/README.md:
```diff
@@ -0,0 +1,235 @@
+
+# Create a business network and manage it
+
+BNE provides a set of workflows that allows a user to start a business network, on-board members and assign them to membership lists or groups. The flows can also be used to update the information
```
What does BNE stand for? Business Network Extension?

tubibuto replied:
Business Networks Extension

benpesterr3 commented on business-networks/README.md:
```diff
@@ -0,0 +1,235 @@
+
+## Amend a membership
+
+There are several ways in which a member's information can be updated, not including network operations such as membership suspension or revocation. These attributes which can be amended are:
```
In the public facing docs, if not here, we should specify how many - and document them all. Several is a bit vague. Fine for now though.

bpaunescu replied:
What I meant here was that "the several ways" of altering the information are updating the listed attributes.

bpaunescu commented:
@benpesterr3 Have addressed your comments and moved the document here https://github.com/corda/bn-extension/pull/4. I've changed the name of one of the classes and reflected the update in the readme.md
